### ISSUE TITLE
[ENH] `y` encoding at template level

### ISSUE THREAD

**Issue by geetu040** [sktime/sktime#6828]
**Is your feature request related to a problem? Please describe.**
Estimators in classification module, seem to encode y at each estimator's class rather than at template level.
by encoding `y` I mean converting
`y = ['a', 'b', 'a', 'a', 'c']`
to
`y = [0, 1, 0, 0, 2]`
before feeding this to the model

**Describe the solution you'd like**
Apply encoding at template level

**Comment by fkiraly** [comment]
agree, this seems to be done by each class individually, and there were a number of historical bugs where implementers dealt only with the int or only with the string case.

I would control the behaviour by a tag, and also keep note of various complications:

* we are dealing with different mtypes, including possibly multivariate `y`, so we need to do an "if pandas" check
* we need to invert the encoding after `predict`

I do not think this can be managed on the level of mtypes, one would need a concept of "sub-mtypes" for this. FYI @julian-fong, @Abhay-Lejith, @pranavvp16, since on an abstract level there is some connection to the boilerplate locations you are looking at.
